File: src/main/java/cn/xihan/http/request/DownloadRequest.java
package cn.xihan.http.request;


import java.io.File;

import cn.xihan.http.EasyLog;
import cn.xihan.http.EasyUtils;
import cn.xihan.http.callback.DownloadCallback;
import cn.xihan.http.config.DownloadApi;
import cn.xihan.http.config.RequestServer;
import cn.xihan.http.listener.OnDownloadListener;
import cn.xihan.http.listener.OnHttpListener;
import cn.xihan.http.model.*;
import okhttp3.Request;

/**
 *    author : Android 轮子哥
 *    github : <a href="https://github.com/getActivity/EasyHttp">https://github.com/getActivity/EasyHttp</a>
 *    time   : 2019/07/20
 *    desc   : 下载请求
 */
public final class DownloadRequest extends HttpRequest<DownloadRequest> {

    private HttpRequest<?> mRealRequest;

    /** 下载请求方式 */
    private HttpMethod mMethod = HttpMethod.GET;

    /** 保存的文件 */
    private File mFile;

    /** 校验的 md5 */
    private String mMd5;

    /** 下载监听回调 */
    private OnDownloadListener mListener;

    /** 请求执行对象 */
    private CallProxy mCallProxy;

    public DownloadRequest() {
        mRealRequest = new GetRequest();
    }

    /**
     * 设置请求方式
     */
    public DownloadRequest method(HttpMethod method) {
        mMethod = method;
        switch (mMethod) {
            case GET:
                // 如果这个下载请求方式是 Get
                mRealRequest = new GetRequest();
                break;
            case POST:
                // 如果这个下载请求方式是 Post
                mRealRequest = new PostRequest();
                break;
            default:
                throw new IllegalStateException("method nonsupport");
        }
        return this;
    }

    /**
     * 设置下载地址
     */
    public DownloadRequest url(String url) {
        server(new RequestServer(url));
        api(new DownloadApi(""));
        return this;
    }

    /**
     * 设置保存的路径
     */

    public DownloadRequest file(String filePath) {
        return file(new File(filePath));
    }

    public DownloadRequest file(File file) {
        mFile = file;
        return this;
    }

    /**
     * 设置 MD5 值
     */
    public DownloadRequest md5(String md5) {
        mMd5 = md5;
        return this;
    }

    /**
     * 设置下载监听
     */
    public DownloadRequest listener(OnDownloadListener listener) {
        mListener = listener;
        return this;
    }

    /**
     * 开始下载
     */
    public DownloadRequest start() {
        long delayMillis = getDelayMillis();
        if (delayMillis > 0) {
            // 打印请求延迟时间
            EasyLog.printKeyValue(this, "RequestDelay", String.valueOf(delayMillis));
        }

        StackTraceElement[] stackTrace = new Throwable().getStackTrace();

        EasyUtils.postDelayed(() -> {

            EasyLog.printStackTrace(this, stackTrace);
            mCallProxy = new CallProxy(createCall());
            new DownloadCallback(this)
                    .setFile(mFile)
                    .setMd5(mMd5)
                    .setListener(mListener)
                    .setCall(mCallProxy)
                    .start();
        }, delayMillis);

        return this;
    }

    /**
     * 取消下载
     */
    public DownloadRequest stop() {
        if (mCallProxy != null) {
            mCallProxy.cancel();
        }
        return this;
    }

    @Override
    public void request(OnHttpListener<?> listener) {
        // 请调用 start 方法
        throw new IllegalStateException("Call the start method");
    }

    @Override
    public <Bean> Bean execute(ResponseClass<Bean> responseClass) {
        // 请调用 start 方法
        throw new IllegalStateException("Call the start method");
    }

    @Override
    public DownloadRequest cancel() {
        // 请调用 stop 方法
        throw new IllegalStateException("Call the start method");
    }

    @Override
    public String getRequestMethod() {
        return String.valueOf(mMethod);
    }

    @Override
    protected Request createRequest(String url, String tag, HttpParams params, HttpHeaders headers, BodyType type) {
        return mRealRequest.api(getRequestApi()).createRequest(url, tag, params, headers, type);
    }

    @Override
    protected void addHttpParams(HttpParams params, String key, Object value, BodyType type) {}

    @Override
    protected void addRequestParams(Request.Builder requestBuilder, HttpParams params, BodyType type) {}

    @Override
    protected void printRequestLog(Request request, HttpParams params, HttpHeaders headers, BodyType type) {}
}

File: src/main/java/cn/xihan/lib/easyhttp/request/DownloadRequest.kt
package cn.xihan.lib.easyhttp.request

import cn.xihan.lib.easyhttp.EasyLog.printKeyValue
import cn.xihan.lib.easyhttp.EasyLog.printStackTrace
import cn.xihan.lib.easyhttp.EasyUtils.postDelayed
import cn.xihan.lib.easyhttp.callback.DownloadCallback
import cn.xihan.lib.easyhttp.config.DownloadApi
import cn.xihan.lib.easyhttp.config.RequestServer
import cn.xihan.lib.easyhttp.listener.OnDownloadListener
import cn.xihan.lib.easyhttp.listener.OnHttpListener
import cn.xihan.lib.easyhttp.model.*
import okhttp3.*
import java.io.File

/**
 * author : Android 轮子哥
 * github : [https://github.com/getActivity/EasyHttp](https://github.com/getActivity/EasyHttp)
 * time   : 2019/07/20
 * desc   : 下载请求
 */
class DownloadRequest : HttpRequest<DownloadRequest>() {
    private var mRealRequest: HttpRequest<*>

    /** 下载请求方式  */
    private var mMethod = HttpMethod.GET

    /** 保存的文件  */
    private var mFile: File? = null

    /** 校验的 md5  */
    private var mMd5: String? = null

    /** 下载监听回调  */
    private var mListener: OnDownloadListener? = null

    /** 请求执行对象  */
    private var mCallProxy: CallProxy? = null

    init {
        mRealRequest = GetRequest()
    }

    /**
     * 设置请求方式
     */
    fun method(method: HttpMethod): DownloadRequest {
        mMethod = method
        mRealRequest = when (mMethod) {
            HttpMethod.GET ->                 // 如果这个下载请求方式是 Get
                GetRequest()
            HttpMethod.POST ->                 // 如果这个下载请求方式是 Post
                PostRequest()
            else -> throw IllegalStateException("method nonsupport")
        }
        return this
    }

    /**
     * 设置下载地址
     */
    fun url(url: String?): DownloadRequest {
        server(RequestServer(url!!))
        api(DownloadApi(""))
        return this
    }

    /**
     * 设置保存的路径
     */
    fun file(filePath: String?): DownloadRequest {
        return file(File(filePath))
    }

    fun file(file: File?): DownloadRequest {
        mFile = file
        return this
    }

    /**
     * 设置 MD5 值
     */
    fun md5(md5: String?): DownloadRequest {
        mMd5 = md5
        return this
    }

    /**
     * 设置下载监听
     */
    fun listener(listener: OnDownloadListener?): DownloadRequest {
        mListener = listener
        return this
    }

    /**
     * 开始下载
     */
    fun start(): DownloadRequest {
        val delayMillis = delayMillis
        if (delayMillis > 0) {
            // 打印请求延迟时间
            printKeyValue(this, "RequestDelay", delayMillis.toString())
        }
        val stackTrace = Throwable().stackTrace
        postDelayed({
            printStackTrace(this, stackTrace)
            mCallProxy = CallProxy(createCall())
            DownloadCallback(this)
                .setFile(mFile)
                .setMd5(mMd5)
                .setListener(mListener)
                .setCall(mCallProxy)
                .start()
        }, delayMillis)
        return this
    }

    /**
     * 取消下载
     */
    fun stop(): DownloadRequest {
        if (mCallProxy != null) {
            mCallProxy!!.cancel()
        }
        return this
    }

    override fun request(listener: OnHttpListener<*>?) {
        // 请调用 start 方法
        throw IllegalStateException("Call the start method")
    }

    override fun <Bean> execute(responseClass: ResponseClass<Bean>?): Bean? {
        // 请调用 start 方法
        throw IllegalStateException("Call the start method")
    }

    override fun cancel(): DownloadRequest {
        // 请调用 stop 方法
        throw IllegalStateException("Call the start method")
    }

    override val requestMethod: String
        get() = mMethod.toString()

    override fun addHttpParams(params: HttpParams, key: String?, value: Any?, type: BodyType?) {

    }

    override fun addRequestParams(requestBuilder: Request.Builder, params: HttpParams, type: BodyType?) {

    }

    override fun printRequestLog(request: Request, params: HttpParams, headers: HttpHeaders, type: BodyType?) {
       
    }


}